Add mutated state when inserting an already existing component

dallenng opened the pull request:
Fixes #333

## Proposed Change

When calling `Archetype::put_dynamic` with `added = false`, the mutated state will be set to true. Because if a component is not added, that means it is updated so it should be marked as mutated.
I added tests to check if the behaviour is correct based on what I think the code should do, feel free to correct me if I'm wrong.

One case is not handled correctly in this PR, but I'll need a bit of guidance if this behaviour is wanted. When the entity contains A and you insert A and B, A is updated and B is added. So, A state should be mutated and B state should be added, but both state are added.

cart commented:
Isn't this behavior already captured by the `Changed<T>` pointer? I suppose theres an argument to be made that we could merge `Changed<T>` and `Mutated<T>` pointers. What we lose is the ability to run logic _only_ when an existing item has been mutated.

dallenng commented:
`Changed<T>` is just a logical OR between `Added<T>` and `Mutated<T>` but the similarity in the name of `Changed<T>` and `Mutated<T>` can easily bring confusion if you're not careful. So, yes this behavior is captured by `Changed<T>`, but it is for the "wrong" reason : triggered by `Added<T>` and not `Mutated<T>`.

Here's a little diagram of what's happening (assuming the commands are run sequentially on the same entity and the trackers are cleared between each command) :

|      Command      |   Entity  |     State in master    |       State in PR      |      Expected state      |
|:-----------------:|:---------:|:----------------------:|:----------------------:|:------------------------:|
|    spawn(A(0))    |   (A(0))  |       `Added<A>`       |       `Added<A>`       |        `Added<A>`        |
|    insert(A(1))   |   (A(1))  |         `None`         |      `Mutated<A>`      |       `Mutated<A>`       |
| insert((A(0), B)) | (A(0), B) | `(Added<A>, Added<B>)` | `(Added<A>, Added<B>)` | `(Mutated<A>, Added<B>)` |

dallenng commented:
As far as I can see, I fixed changed tracking on `Commands::insert` and this PR is ready to merge unless I have to add an entry to the changelog or something else.

cart commented on crates/bevy_ecs/hecs/src/archetype.rs:
```diff
@@ -400,6 +400,8 @@ impl Archetype {
         let state = self.state.get_mut(&ty).unwrap();
         if added {
             state.added_entities[index] = true;
+        } else {
```
Hey sorry for the delay here. These changes make sense to me. I _think_ this works as-is, but I'm a little worried that the implicit `!added == mutated` here could cause problems / create corner case issues. I'd rather just pass in a `mutated` flag and set it explicitly each time.

cart replied:
Im specifically worried about cases where we change archetypes and some of the components aren't mutated.

dallenng replied:
Ok, I did the changes you suggested and I think it's better like this because further call of the method will need to be explicit on the state.

cart commented:
Sorry for dragging my feat on this one. This looks good to go. Thanks for sticking with this / improving our correctness!
